### PR TITLE
Add interceptor logic to allow custom conceptScheme fetching

### PR DESCRIPTION
**Pull request by x-m-el**
## Description
Add logic to allow adding a `interceptors.ts` file to the config that will get called when fetching conceptSchemes defined in the form. This allows populating dropdown fields (like `displayTypes:conceptSchemeSelector`) with any (possibly dynamic) data in the triplestore, without needing to make it an actual concept. It also allows filtering out conceptScheme triples that form-content-service fetched, giving extra flexibility.
If not defined, the interceptor does nothing, so this is a non-breaking change.

In the future it would be nice to force forms to be put in a subfolder (e.g. /forms) in the config directory, so other configurations can be added more easily.

It wasn't clear to me where in the Readme this should be explained.
## How to test
Create a form with some dropdown and a custom uri for the concept scheme. Code that will fill it with all `foaf:Person`s in the database as an example:

```
export async function interceptorGetConceptSchemeTriples(
  conceptSchemeUrisList,
  fetchedSpoBindings,
) {
  const conceptSchemeUri = "http://redpencil.data.gift/concept-schemes/CustomConceptPerson"
  if(conceptSchemeUrisList.includes(conceptSchemeUri)) {
    const result = await query(`
      PREFIX foaf: <http://xmlns.com/foaf/0.1/>
      SELECT DISTINCT ?uri ?name WHERE {
        ?uri a foaf:Person .
        ?uri foaf:name ?name .   }
    `);
    const concepts = result.results.bindings.map(({uri, name}) => ({uri: uri.value, label: name.value}))
    const conceptBindings = triplesForScheme({conceptSchemeUri, conceptSchemeLabel: "Persons", concepts})
    return fetchedSpoBindings.concat(conceptBindings);
  } else {
  return fetchedSpoBindings;  
}

function toSPO(s, p, o, typeO = "uri") {
  return {
    "s": { "type": "uri", "value": s },
    "p": { "type": "uri", "value": p },
    "o": { "type": typeO === "uri"? typeO : "literal", "value": o }
  }
}

function triplesForScheme({conceptSchemeUri, conceptSchemeLabel, concepts /* list of { uri, label }*/}) {
  const bindings = [
    toSPO(conceptSchemeUri, "http://www.w3.org/1999/02/22-rdf-syntax-ns#type", "http://www.w3.org/2004/02/skos/core#Concept"),
    toSPO(conceptSchemeUri, "skos:prefLabel", conceptSchemeLabel, "string")
  ]
  for(const {uri, label} of concepts) {
    bindings.push(toSPO(uri, "http://www.w3.org/1999/02/22-rdf-syntax-ns#type", "http://www.w3.org/2004/02/skos/core#Concept"))
    bindings.push(toSPO(uri, "http://www.w3.org/2004/02/skos/core#prefLabel", label, "string"))
    bindings.push(toSPO(uri, "http://www.w3.org/2004/02/skos/core#inScheme", conceptSchemeUri))
  }
  return bindings;
}
```

### CODE REVIEW

**Review by cecemel — COMMENTED** [review]
Hi,

Thanks for the PR. I think we can grind with this for our use case. And it works!

However, in my head, I thought of a more aggressive escape hatch, not limited to concept schemes but allowing maximum manipulation of the data before sending it to the frontend.

So, a more appropriate place would be:
  [services/forms-from-config.ts#L31](https://github.com/lblod/form-content-service/blob/9b2a0a11fb5b1d9863808e47fde175b4983ba781/services/forms-from-config.ts#L31)

Having a function like `interceptorMeta(formTTL)` that appends whatever triples it needs to find based on the form-defintion.

Thinking about it, over time, even this approach might not sufficient:
  - We've encountered use cases in the past where we had a custom component that validated input based on the population of an administrative unit.
    To fetch the correct information, we also had to know who was making the request.
      See e.g. https://github.com/lblod/app-subsidiepunt/blob/3053cf3c7aeb5b1a44a7723148559a8b244e2ba2/config/subsidy-application-management/forms/climate/step-submit-proposal/tailored/meta/extractors/climate-subsidy-extractor.js#L9  which had a bit of the same pattern. (Don't look at the implementation, old piece of code)


In any case, that's a bit of general rambling.

Concrete action points would be:

 - Flag it as a draft here
 - I added you as maintainer of the repo, so you don't need to work with a fork
 - Can you give this branch some experimental tag `v0.0.8-experiment_1`. Build system should pick this up. And then you can include it PR of `app-topology-change-context-poc`
 - The eternal serialization issue of triples, isn't there some util in the code we could use in the interceptor? Now we have to do
    ```
    function toSPO(s, p, o, typeO = "uri") {
  return {
    "s": { "type": "uri", "value": s },
    "p": { "type": "uri", "value": p },
    "o": { "type": typeO === "uri"? typeO : "literal", "value": o }
  }
    ```

thanks

**Comment by x-m-el** [comment]
You could have a general Meta interceptor, and if needed I would think it would be nice to add this, but that adds extra overhead to maintaining your interceptor, e.g. if the forms definitions change (which can happen pretty quickly: there are currrently 2 useable form definitions to define the used concept scheme, of which only the "options" one works for now, so it might already be needed at a next update).
With this specific interceptor, there should be no need to update the interceptor code if forms change, as long as they still handle conceptSchemes. The interceptor is also more contained in what it does and much more easy to understand, versus also needing to know the model of the forms.
Basically this interceptor and a general meta interceptor have different usecases.

as for serialization: yes this is annoying. I tried to find a util function in the service, but couldn't find any. Most are very specific to the service. 
Using the structure of "bindings" felt logical, as this is a known structure from using the `query` function (helping to do filtering if needed) and contains all information to serialize the data afterwards.
